File: src/types/photo.ts
/**
 * 写真データの型定義
 */

/**
 * GPS座標情報
 */
export interface Gps {
  /** 緯度 */
  lat: number
  /** 経度 */
  lng: number
}

/**
 * カメラ情報
 */
export interface CameraInfo {
  /** メーカー名 */
  make: string
  /** モデル名 */
  model: string
}

/**
 * EXIF情報
 */
export interface ExifData {
  /** GPS座標（存在しない場合はnull） */
  gps: Gps | null
  /** 撮影日時（ISO 8601形式、存在しない場合はnull） */
  datetime: string | null
  /** カメラ情報（存在しない場合はnull） */
  camera: CameraInfo | null
  /** 画像の幅（ピクセル） */
  width: number | null
  /** 画像の高さ（ピクセル） */
  height: number | null
  /** ISO感度 */
  iso: number | null
  /** 絞り値（F値） */
  aperture: number | null
  /** シャッター速度（秒） */
  shutterSpeed: number | null
  /** 焦点距離（mm） */
  focalLength: number | null
}

/**
 * 写真データ
 */
export interface PhotoData {
  /** ファイルの絶対パス */
  path: string
  /** ファイル名（拡張子含む） */
  filename: string
  /** ファイルサイズ（バイト） */
  fileSize: number
  /** ファイルの最終更新日時（ISO 8601形式） */
  modifiedTime: string
  /** EXIF情報（存在しない場合はnull） */
  exif: ExifData | null
  /** サムネイル画像（Base64エンコード、未生成の場合はnull） */
  thumbnail: string | null
}

/**
 * 写真リストのフィルター条件
 */
export interface PhotoFilter {
  /** 撮影日時の開始日（この日付以降） */
  dateFrom: Date | null
  /** 撮影日時の終了日（この日付以前） */
  dateTo: Date | null
  /** GPS情報の有無でフィルタリング */
  hasGps: boolean | null
}

/**
 * 地図の表示設定
 */
export interface MapSettings {
  /** 中心座標 */
  center: Gps
  /** ズームレベル（1-18） */
  zoom: number
  /** 地図タイルの種類 */
  tileLayer: 'osm' | 'google' | 'satellite'
}

/**
 * ディレクトリエントリ（フォルダまたはファイル）
 */
export interface DirectoryEntry {
  /** エントリの名前（ファイル名またはフォルダ名） */
  name: string
  /** エントリの絶対パス */
  path: string
  /** ディレクトリかどうか（true: フォルダ, false: ファイル） */
  isDirectory: boolean
  /** 最終更新日時（ISO 8601形式） */
  modifiedTime: string
  /** ファイルサイズ（バイト、ディレクトリの場合は0） */
  fileSize: number
}

/**
 * ディレクトリの内容
 */
export interface DirectoryContent {
  /** 現在のディレクトリパス */
  currentPath: string
  /** 親ディレクトリパス（ルートディレクトリの場合はnull） */
  parentPath: string | null
  /** ディレクトリ内のエントリ一覧（フォルダとファイル） */
  entries: DirectoryEntry[]
}

File: src-tauri/src/models/photo.rs
use serde::{Deserialize, Serialize};

/// GPS座標情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Gps {
    /// 緯度
    pub lat: f64,
    /// 経度
    pub lng: f64,
}

/// カメラ情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CameraInfo {
    /// メーカー名
    pub make: String,
    /// モデル名
    pub model: String,
}

/// EXIF情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExifData {
    /// GPS座標（存在しない場合はNone）
    pub gps: Option<Gps>,
    /// 撮影日時（ISO 8601形式、存在しない場合はNone）
    pub datetime: Option<String>,
    /// カメラ情報（存在しない場合はNone）
    pub camera: Option<CameraInfo>,
    /// 画像の幅（ピクセル）
    pub width: Option<u32>,
    /// 画像の高さ（ピクセル）
    pub height: Option<u32>,
    /// ISO感度
    pub iso: Option<u32>,
    /// 絞り値（F値）
    pub aperture: Option<f64>,
    /// シャッター速度（秒）
    pub shutter_speed: Option<f64>,
    /// 焦点距離（mm）
    pub focal_length: Option<f64>,
}

/// 写真データ
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PhotoData {
    /// ファイルの絶対パス
    pub path: String,
    /// ファイル名（拡張子含む）
    pub filename: String,
    /// ファイルサイズ（バイト）
    pub file_size: u64,
    /// ファイルの最終更新日時（ISO 8601形式）
    pub modified_time: String,
    /// EXIF情報（存在しない場合はNone）
    pub exif: Option<ExifData>,
    /// サムネイル画像（Base64エンコード、未生成の場合はNone）
    pub thumbnail: Option<String>,
}

impl PhotoData {
    /// ファイルパスから PhotoData を作成（EXIF情報とサムネイルは後から追加）
    #[allow(dead_code)]
    pub fn new(path: String, filename: String, file_size: u64, modified_time: String) -> Self {
        Self {
            path,
            filename,
            file_size,
            modified_time,
            exif: None,
            thumbnail: None,
        }
    }

    /// GPS情報を持っているかチェック
    #[allow(dead_code)]
    pub fn has_gps(&self) -> bool {
        self.exif.as_ref().and_then(|e| e.gps.as_ref()).is_some()
    }

    /// 撮影日時を持っているかチェック
    #[allow(dead_code)]
    pub fn has_datetime(&self) -> bool {
        self.exif
            .as_ref()
            .and_then(|e| e.datetime.as_ref())
            .is_some()
    }
}

/// ディレクトリエントリ（フォルダまたはファイル）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DirectoryEntry {
    /// エントリの名前（ファイル名またはフォルダ名）
    pub name: String,
    /// エントリの絶対パス
    pub path: String,
    /// ディレクトリかどうか（true: フォルダ, false: ファイル）
    pub is_directory: bool,
    /// 最終更新日時（ISO 8601形式）
    pub modified_time: String,
    /// 撮影日時（ISO 8601形式、存在しない場合はNone）
    pub captured_time: Option<String>,
    /// ファイルサイズ（バイト、ディレクトリの場合は0）
    pub file_size: u64,
}

/// ディレクトリの内容
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DirectoryContent {
    /// 現在のディレクトリパス
    pub current_path: String,
    /// 親ディレクトリパス（ルートディレクトリの場合はNone）
    pub parent_path: Option<String>,
    /// ディレクトリ内のエントリ一覧（フォルダとファイル）
    pub entries: Vec<DirectoryEntry>,
}
